frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.unit.spec.tsx

```diff
@@ -0,0 +1,228 @@
+import React from "react";
+import _ from "underscore";
+import nock from "nock";
+import userEvent from "@testing-library/user-event";
+import { waitFor } from "@testing-library/react";
+
+import { render, screen } from "__support__/ui";
+
+import {
+  createMockActionParameter,
+  createMockQueryAction,
+  createMockImplicitQueryAction,
+} from "metabase-types/api/mocks";
+
+import ActionParametersInputForm, {
+  ActionParametersInputModal,
+} from "./ActionParametersInputForm";
+
+const defaultProps = {
+  missingParameters: [
+    createMockActionParameter({
+      id: "1",
+      name: "Parameter 1",
+      type: "type/Text",
+    }),
+    createMockActionParameter({
+      id: "2",
+      name: "Parameter 2",
+      type: "type/Text",
+    }),
+  ],
+  dashcardParamValues: {},
+  action: createMockQueryAction(),
+  dashboard: { id: 123 },
+  dashcard: { id: 456 },
+  onSubmit: jest.fn(() => ({ success: true })),
+  onCancel: _.noop,
+  onSubmitSuccess: _.noop,
+};
+
+async function setup(options?: any) {
+  render(<ActionParametersInputForm {...defaultProps} {...options} />);
+}
+
+async function setupModal(options?: any) {
+  render(
+    <ActionParametersInputModal
+      title="Test Modal"
+      onClose={_.noop}
+      {...defaultProps}
+      {...options}
+    />,
+  );
+}
+
+function setupPrefetch() {
+  nock(location.origin)
+    .get(uri => uri.includes("/api/dashboard/123/dashcard/456/execute"))
+    .reply(200, {
+      "1": "uno",
+      "2": "dos",
+    });
+}
+
+describe("Actions > ActionParametersInputForm", () => {
+  afterEach(() => {
+    nock.cleanAll();
+  });
+
+  it("should render an action form", async () => {
+    await setup();
+    expect(screen.getByTestId("action-form")).toBeInTheDocument();
+    expect(screen.getByText("Parameter 1")).toBeInTheDocument();
+    expect(screen.getByText("Parameter 2")).toBeInTheDocument();
+  });
+
+  it("should call onCancel when clicking the cancel button", async () => {
+    const cancelSpy = jest.fn();
+    await setup({ onCancel: cancelSpy });
+    await userEvent.click(screen.getByText("Cancel"));
+    expect(cancelSpy).toHaveBeenCalled();
```

Comment:
Instead of creating mocks per test, what do you think about mocking them inside the `setup` and then return then? Example:

```jsx
function setup({ onCancel = jest.fn() }) {
  render(...)
  return { onCancel }
}

it("should call onCancel", () => {
  const { onCancel } = setup();
  userEvent.click(...)
  expect(onCancel).toHaveBeenCalled();
})
```

Reply:
I don't see a problem with that, I was just thinking that I didn't need to create and return a mock for every test when I'm only using it once.